### PR TITLE
TRI-217/Project Details and Members Endpoint

### PR DESCRIPTION
**Pull request by luccar31**
# Pull Request

## Description
This PR adds an endpoint that allows to retrieve all projects a user is belongs to, the members associated and the roles of each one if the user requesting the data is a project manager. Also, this implements a temporary solution to providing the issue provider name to integration endpoints. Future refactor to fully support this feature is needed.

## Ticket Link
[TRI-217](https://linear.app/tricker-sirius/issue/TRI-217/project-details-and-members-endpoint)

## Steps to Reproduce
1. Run `npm run test`
2. Run `docker compose up`
3. Test the endpoint `api/project/me` with Swagger or Postman.

## Images/Screenshots
image

### CODE REVIEW

File: src/documentation/domains/project/controller/project.controller.docs.ts [hunk 1]
```diff
@@ -1,6 +1,28 @@
 /**
  * @swagger
  * paths:
+ *   /me:
```
Comment (Juan-Bianchi) [inline src/documentation/domains/project/controller/project.controller.docs.ts]: please update doc

File: src/domains/project/service/project.service.impl.ts [hunk 2]
```diff
@@ -1,18 +1,61 @@
...
+   * @returns {Promise<ProjectViewDTO[]>} - Promise resolving to an array of ProjectViewDTO objects representing the user's projects.
+   * @throws {NotFoundException} - Throws an error if the user is not found.
+   */
+  async getMyProjectInfo(userId: string): Promise<ProjectViewDTO[]> {
+    const user = await this.userRepository.getById(userId);
+    if (user == null) throw new NotFoundException('User');
+
+    const views: ProjectViewDTO[] = [];
+
+    const userProjectRoleByUser = await this.userProjectRoleRepository.getUserRolesByUserAndOrProject({ userId });
```
Comment (Juan-Bianchi) [inline src/domains/project/service/project.service.impl.ts]: Please add type here, is it an array or an object?
Why do you retrieve all projects and roles of the user? Shouldn't we only retrieve data for an specific project?
I believe  when user selects the project in UI, frontend request the data of that specific project.

File: src/domains/integration/controller/integration.controller.ts [hunk 3]
```diff
@@ -33,18 +36,28 @@ const organizationRepository: OrganizationRepository = new OrganizationRepositor
 const emailService: EmailService = new MailgunEmailService(mailgunClient);
 const administratorRepository: AdministratorRepository = new AdministratorRepositoryImpl(db);
 const integrationRepository: IntegrationRepository = new IntegrationRepositoryImpl(db);
-const service: IntegrationService = new IntegrationServiceImpl(adapter, projectRepository, userRepository, pendingAuthRepository, pendingMemberMailsRepository, organizationRepository, administratorRepository, integrationRepository, emailService);
-
-integrationRouter.post('/linear/projects', withAwsAuth, validateRequest(ProviderKeyDTO, 'body'), async (req: Request, res: Response) => {
-  const { key, provider } = req.body as unknown as ProviderKeyDTO;
+const issueProviderRepository: IssueProviderRepository = new IssueProviderRepositoryImpl(db);
+const service: IntegrationService = new IntegrationServiceImpl(adapter, projectRepository, userRepository, pendingAuthRepository, pendingMemberMailsRepository, organizationRepository, administratorRepository, integrationRepository, emailService, issueProviderRepository);
+
```
Comment (Juan-Bianchi) [inline src/domains/integration/controller/integration.controller.ts]: Please, move the DTO to a dto folder and document the class

File: src/domains/integration/controller/integration.controller.ts [hunk 4]
```diff
@@ -33,18 +36,28 @@ const organizationRepository: OrganizationRepository = new OrganizationRepositor
...
+export class IssueProviderURLParams {
+  @IsString()
+  @IsNotEmpty()
+  @IsValidIssueProvider()
+  @Transform(({ value }) => value.trim().toUpperCase())
+  issueProviderName!: string;
+}
+
```
Comment (Juan-Bianchi) [inline src/domains/integration/controller/integration.controller.ts]: Please remember to add all endpoints new param to Swagger documentation

File: src/domains/issueProvider/repository/IssueProvider.repository.impl.ts [hunk 5]
```diff
@@ -5,6 +5,16 @@ import type { IssueProvider, PrismaClient } from '@prisma/client';
 export class IssueProviderRepositoryImpl implements IssueProviderRepository {
   constructor(private readonly db: PrismaClient) {}
 
+  async getById(id: string): Promise<IssueProviderDTO | null> {
```
Comment (Juan-Bianchi) [inline src/domains/issueProvider/repository/IssueProvider.repository.impl.ts:8]: Please add doc

File: src/domains/project/service/project.service.ts [hunk 6]
```diff
@@ -1,7 +1,13 @@
-import { type DevProjectFiltersDTO, type PMProjectFiltersDTO } from '@domains/project/dto';
+import { type DevProjectFiltersDTO, type PMProjectFiltersDTO, type MyProjectViewDTO } from '@domains/project/dto';
 import { type UserProject } from '@domains/issue';
```
Comment (Juan-Bianchi) [inline src/domains/project/service/project.service.ts:3]: Please put interfaces in dto folder and document them

**Review by Juan-Bianchi — CHANGES_REQUESTED** [review]
Please update documentation

File: src/domains/project/repository/project.repository.impl.ts [hunk 7]
```diff
@@ -15,25 +24,84 @@ export class ProjectRepositoryImpl implements ProjectRepository {
...
+              some: {
+                userId,
+              },
+            },
+          },
+        ],
+      },
+      include: {
+        usersRoles: {
+          select: {},
```
Comment (Juan-Bianchi) [inline src/domains/project/repository/project.repository.impl.ts]: Please correct the query, you cant add select and include both... I attach the error message returned when the endpoint was tested with swagger
image
image

File: src/domains/integration/controller/integration.controller.ts [hunk 8]
```diff
@@ -35,16 +34,17 @@ const administratorRepository: AdministratorRepository = new AdministratorReposi
 const integrationRepository: IntegrationRepository = new IntegrationRepositoryImpl(db);
 const service: IntegrationService = new IntegrationServiceImpl(adapter, projectRepository, userRepository, pendingAuthRepository, pendingMemberMailsRepository, organizationRepository, administratorRepository, integrationRepository, emailService);
 
-integrationRouter.post('/linear/projects', validateRequest(ProviderKeyDTO, 'body'), async (req: Request, res: Response) => {
-  const { key, provider } = req.body as unknown as ProviderKeyDTO;
+integrationRouter.post('/:issueProviderName/projects', validateRequest(IssueProviderURLParams, 'params'), validateRequest(ApiKeyDTO, 'body'), async (req: Request<any, any, ApiKeyDTO>, res: Response) => {
```
Comment (Juan-Bianchi) [inline src/domains/integration/controller/integration.controller.ts:37]: Frontend is sending an object with two different properties in the body of the request and ApiKeyDTO only accepts one. It is triggering a validation exception from the validation class middleware. Please correct our DTO or contact frontend team so they fix what they are sending

File: src/domains/project/service/project.service.impl.ts [hunk 9]
```diff
@@ -13,6 +14,43 @@ export class ProjectServiceImpl implements ProjectService {
...
+   * @returns {Promise<MyProjectViewDTO>} - Promise resolving to a ProjectViewDTO object with project details.
+   * @throws {NotFoundException} - Throws NotFoundException with specific messages:
+   *                                 * 'User': If the user with the provided ID is not found.
+   *                                 * 'Project': If the project with the provided ID is not found or the user doesn't have access to it.
+   */
+  async getMyProjectView({ userId, projectId }: GetProjectParams): Promise<MyProjectViewDTO> {
+    const user = await this.userRepository.getByCognitoId(userId);
+    if (user == null) throw new NotFoundException('User');
+
+    const project: ProjectDetailsDTO | null = await this.projectRepository.getProjectDetails(userId, projectId);
```
Comment (Juan-Bianchi) [inline src/domains/project/service/project.service.impl.ts]: projectId sent to this method is actually cognitoId. As in the repository you search by tricker userId, user.id should be sent as argument.  In addition order of arguments must be inverted, method signature receives project id and then user id.

File: src/domains/project/service/project.service.impl.ts [hunk 10]
```diff
@@ -13,6 +14,43 @@ export class ProjectServiceImpl implements ProjectService {
...
+    if (project == null) throw new NotFoundException('Project');
+
+    const pmUserAndRole: UserRoleDTO | undefined = project.usersAndRoles.find((userAndRole) => userAndRole.role.name === ProjectRole.PROJECT_MANAGER);
```
Comment (Juan-Bianchi) [inline src/domains/project/service/project.service.impl.ts:35]: check this method, is not working ok

**Review by Juan-Bianchi — CHANGES_REQUESTED** [review]
Please fix logic errors

File: src/domains/project/service/project.service.impl.ts (hunk 10 above)
Comment (luccar31) [reply src/domains/project/service/project.service.impl.ts:35]: image

working now

File: src/domains/project/service/project.service.impl.ts (hunk 9 above)
Comment (luccar31) [reply src/domains/project/service/project.service.impl.ts]: done

**Review by Juan-Bianchi — APPROVED** [review]

**Pull request merged**